### PR TITLE
Added event to prevent tainting.

### PR DESCRIPTION
**Pull request by mortenson**
Per https://github.com/vimeo/psalm/issues/5382 this PR adds a new event that allows plugins to prevent tainting in specific scenarios. I've also added an example plugin/test that shows how you can inspect the structure of an array to determine if it should be considered a taint source.

This is a WIP, so let me know if you want the event name or test changed - once that's decided I can dispatch the event from more analyzers if you'd like.

### CODE REVIEW

File: src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php [hunk 1]
```diff
@@ -220,7 +224,8 @@ private static function analyzeArrayItem(
         Context $context,
         ArrayCreationInfo $array_creation_info,
         int $int_offset,
-        PhpParser\Node\Expr\ArrayItem $item
+        PhpParser\Node\Expr\ArrayItem $item,
+        Codebase $codebase = null
```
Comment (muglug) [inline src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php:228]: this is never not null

Comment (mortenson) [reply src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php:228]: I was thinking of backwards compatibility for changing the method signature but I see it's private now.

Comment (mortenson) [reply src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php]: Fixed

**Comment by muglug** [comment]
Ok, extra complication: I think this plugin should instead inform these two arguments: https://github.com/vimeo/psalm/blob/dd4d970a37a1a4dde56f7903333e2ef3c5481394/src/Psalm/Internal/Codebase/DataFlowGraph.php#L30-L31 in the call here: https://github.com/vimeo/psalm/blob/df43126626715cfe4e641c2017a6b41a8a43c433/src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php#L371

It may be the case that you want to remove the `html` taint but preserve the `sql` taint, for example. In this case you'd want to remove `html` taints.

So instead of being a boolean I think it wants to provide two separate arrays of added and removed taints at every node it's used on. If both those arrays are empty there's no net effect

**Comment by muglug** [comment]
I've updated the PR along those lines! The main benefit is it allows people to institute their own custom taints in a smarter fashion.

**Comment by mortenson** [comment]
@muglug That looks good to me - the test I added still works as expected so I think this would give me a path forward on the render array problem. Would you like me to rename the new test method and add this event dispatch to other analyzers? (If you have time to do this don't wait on me though)

**Comment by muglug** [comment]
> Would you like me to rename the new test method and add this event dispatch to other analyzers?

That would be great!

**Comment by mortenson** [comment]
@muglug StaticCallAnalyzer, AtomicPropertyFetchAnalyzer, and ArrayFetchAnalyzer didn't have access to Context which is needed for this event, should I add $context as an optional parameter to the relevant public static methods on those classes?

All other analyzers that used that `TaintedInput` check you mentioned in the issue have been updated.

**Comment by mortenson** [comment]
I did the above, feel free to clean up the new event invocations/calls as needed.

**Comment by muglug** [comment]
Thanks, this is a great effort

**Comment by mortenson** [comment]
Thank you @muglug !

**Pull request merged**